c++/src/kj/string.h

```diff
@@ -288,8 +289,76 @@ String heapString(ArrayPtr<const char> value);
 // Magic str() function which transforms parameters to text and concatenates them into one big
 // String.
 
-namespace _ {  // private
+template <typename T>
+class HasSizeMemberFunction_ {
+  using yes = char;
+  using no = yes[2];
+
+  template<typename C> static auto test(void*) -> decltype(instance<C>().size(), yes{});
+  template<typename C> static no& test(...);
+
+public:
+
+  static constexpr bool value = sizeof(test<const T&>(nullptr)) == sizeof(yes) ||
+      sizeof(test<T&&>(nullptr)) == sizeof(yes);
+};
+
+template <typename T>
+class HasBeginMemberFunction_ {
+  using yes = char;
+  using no = yes[2];
 
+  template<typename C> static auto test(void*) -> decltype(instance<C>().begin(), yes{});
+  template<typename C> static no& test(...);
+
+public:
+
+  static constexpr bool value = sizeof(test<const T&>(nullptr)) == sizeof(yes) ||
+      sizeof(test<T&&>(nullptr)) == sizeof(yes);
+};
+
+template <typename T, typename = void>
+class SignalSafeCharSequence: public T {
+  // Annotates that the result of a stringification operation is signal-safe. T is the type being
+  // wrapped after stringification (ArrayPtr/CappedArray/FixedArray/StringPtr).
+public:
+  using T::T;
+
+  SignalSafeCharSequence() = default;
+  template <typename ...Args>
+  constexpr SignalSafeCharSequence(Args&&... args): T(kj::fwd<Args>(args)...) {}
+};
+template <typename T>
+class SignalSafeCharSequence<T,
+    EnableIf<HasSizeMemberFunction_<T>::value && !HasBeginMemberFunction_<T>::value>>: public T {
```

Comment:
> On another note, though, it seems like this pattern matching isn't accomplishing anything? toCharSequence() is required to return a type that has begin() and size() methods. Why even detect and handle any other case, when they won't work anyway

Not everything is possible to supply to `toCharSequence`.

> Not necessarily asking you to change it, but this really confused me.

Yeah, I couldn't think of a simpler way as I need SFINAE because of the reason above. I'm not sure your suggestion would work either but something I'll try later in a cleanup PR in the future. The `EnableIf` trick wasn't my idea - it's a pretty common pattern on the web. It's also why `EnableIf` takes a type rather than doing `void` - if ours didn't have `void` I think I could get rid of the dummy second template argument.

Reply:
> Not everything is possible to supply to toCharSequence.

I don't understand your point.

When does `SignalSafeCharSequence` ever wrap a type that doesn't have `begin()` and `size()` methods?